sweep/sweeper.go

```diff
@@ -29,13 +64,387 @@ type UtxoSweeperConfig struct {
+
+	// ChainIO allows us to query the state of the current main chain.
+	ChainIO lnwallet.BlockChainIO
+
+	HasSpendingTx func(op wire.OutPoint) (*wire.MsgTx, error)
 }
 
-// New returns a new UtxoSweeper instance.
+// Result is the struct that is pushed through the result channel. Callers
+// can use this to be informed of the final sweep result. In case of a remote
+// spend, ErrRemoteSpend is return in err.
+type Result struct {
+	Err error
+	Tx  *wire.MsgTx
+}
+
+type sweeperInput struct {
+	input      Input
+	resultChan chan Result
+}
+
+type pendingInput struct {
+	txHash        chainhash.Hash
+	listeners     []chan Result
+	ntfnRegCancel func()
+}
+
+// New returns a new Sweeper instance.
 func New(cfg *UtxoSweeperConfig) *UtxoSweeper {
 	return &UtxoSweeper{
-		cfg: cfg,
+		cfg:           cfg,
+		newInputs:     make(chan *sweeperInput),
+		spendChan:     make(chan *chainntnfs.SpendDetail),
+		quit:          make(chan struct{}),
+		pendingInputs: make(map[wire.OutPoint]*pendingInput),
+	}
+}
+
+// Start starts the process of constructing and publish sweep txes.
+func (s *UtxoSweeper) Start() error {
+
+	if !atomic.CompareAndSwapUint32(&s.started, 0, 1) {
+		return nil
+	}
+
+	log.Debugf("Starting sweeper")
+
+	s.wg.Add(1)
+	go s.collector()
+
+	return nil
+}
+
+// Stop stops sweeper from listening to block epochs and constructing sweep
+// txes.
+func (s *UtxoSweeper) Stop() error {
+	if !atomic.CompareAndSwapUint32(&s.stopped, 0, 1) {
+		return nil
+	}
+
+	log.Infof("Sweeper shutting down")
+
+	close(s.quit)
+	s.wg.Wait()
+
+	log.Infof("Sweeper shut down")
+
+	return nil
+}
+
+// SweepInput sweeps inputs back into the wallet. The inputs will be batched and
+// swept after the batch time window ends.
+func (s *UtxoSweeper) SweepInput(input Input) (chan Result, error) {
+	log.Debug("Pushing new input to collector")
+
+	sweeperInput := &sweeperInput{
+		input:      input,
+		resultChan: make(chan Result),
+	}
+
+	// Deliver input to main event loop.
+	select {
+	case s.newInputs <- sweeperInput:
+	case <-s.quit:
+		return nil, fmt.Errorf("sweeper shutting down")
+	}
+
+	return sweeperInput.resultChan, nil
+}
+
+func (s *UtxoSweeper) collector() {
+	log.Debug("Collector started")
+	defer s.wg.Done()
+
+	s.cfg.Ticker.Resume()
+	defer s.cfg.Ticker.Stop()
+
+	inputToSweep := make(map[wire.OutPoint]Input)
+	clearInputsToSweep := func() {
+		inputToSweep = make(map[wire.OutPoint]Input)
+	}
+	sweepCountdown := 0
+	for {
+		select {
+		case input := <-s.newInputs:
+			outpoint := *input.input.OutPoint()
+			pendInput, pending := s.pendingInputs[outpoint]
+			if pending {
+				log.Debugf("Already pending input %v received",
+					outpoint)
+
+				// Add additional result channel to signal spend
+				// of this input.
+				pendInput.listeners =
+					append(pendInput.listeners,
+						input.resultChan)
+
+				break
+			}
+
+			// Lookup if there is already a confirmed or unconfirmed
+			// tx of ours that spends this input.
+			spendingTx, err :=
```

Comment:
Adhering to 80 char columns, I think this should be able to be placed on the line above.

Reply:
The comment you mean? In my editor with tab size 8, the top line is exactly 80 chars already? The `signalResultChan` call below is too long, I split that.